lib/TInitialState.php

```diff
@@ -0,0 +1,125 @@
+<?php
+declare(strict_types=1);
+/**
+ * @copyright Copyright (c) 2020 Joas Schilling <coding@schilljs.com>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Talk;
+
+
+use OC\HintException;
+use OC\User\NoUserException;
+use OCP\App\IAppManager;
+use OCP\Files\IRootFolder;
+use OCP\Files\NotPermittedException;
+use OCP\ICacheFactory;
+use OCP\IConfig;
+use OCP\IInitialStateService;
+use OCP\IUser;
+use OCP\Util;
+
+trait TInitialState {
+
+	/** @var Config */
+	protected $talkConfig;
+	/** @var IConfig */
+	protected $serverConfig;
+	/** @var IInitialStateService */
+	protected $initialStateService;
+	/** @var ICacheFactory */
+	protected $memcacheFactory;
+
+	protected function publishInitialStateShared(): void {
+		// Needed to enable the screensharing extension in Chromium < 72.
+		Util::addHeader('meta', ['id' => 'app', 'class' => 'nc-enable-screensharing-extension']);
+
+		$this->initialStateService->provideInitialState(
+			'talk', 'prefer_h264',
+			$this->serverConfig->getAppValue('spreed', 'prefer_h264', 'no') === 'yes'
+		);
+
+		$signalingMode = $this->talkConfig->getSignalingMode();
+		if ($signalingMode !== 'internal' && !$this->memcacheFactory->isAvailable()) {
```

Comment:
Is a distributed cache always needed for external signaling, or only for clustered external signaling?

In any case, I think this is too agressive. If an admin sets two external signaling servers and then happily navigates away from the settings to the Files app the admin will be greeted with that message. But then it will not be possible to open again the settings (at least, without using the direct URL) as trying to open the root Nextcloud URL will redirect to the Files app and the message will be shown again.

On the other hand, if setting several external signaling servers were prevented in the settings if a distributed cache is not set that would work when setting the servers, but if the distributed cache is then removed... That would be a problem. I am not sure how to handle that. So... I guess that the `HintException` is OKish :shrug: Yes, I am thinking out loud :-P

Reply:
> Is a distributed cache always needed for external signaling, or only for clustered external signaling?

So it's not needed for the normal external signaling, but I wouldn't run a Nextcloud without it anyway if I need the HPB. But I added an appconfig now `signaling_dev` to not overwrite the signaling mode and not require the memcache